docs/rules/no-unused-expressions.md

```diff
@@ -59,6 +59,8 @@ delete a.b
 void a
 ```
 
+Strict mode directives are also not considered problems.
```

Comment:
As far as all directives being permitted, yes, @michaelficarra, I myself also would favor the more generic wording presented by @pedrottimark.

(Incidentally, however, I think a new issue should be created to flag directives except strict mode and those whitelisted, such as "use asm". Otherwise, it is possible someone may mistype and get unexpected code behaviors or fail extra validation. While the "strict" rule covers this for strict mode, maybe this rule should be merged into such a proposed general purpose one. I haven't looked through the code carefully, but it also looks like the code may fail to block usage of EscapeSequence and LineContinuation which are not permitted for a strict directives.)

As far as it being "obvious", @michaelficarra, that directives are not expressions, while this may be true to those who have read the spec, presumably this tool is intended to help everyone, including those who are not familiar with every bit of arcana in the spec (how often do "directives" as a category come up, anyways? Yes strict mode is common, but one might assume it (and directives) are just hacks used by ES engines rather than features explicitly anticipated in the language, and indeed they were before ES5). For those who haven't read the relevant parts of the modern specs, a string expression statement at the top of a module, script, or function, is indistinguishable from an unused expression (statement). There is no need to explicitly mention different kinds of statements and declarations as non-problematic because it is clear to those with some background in the language that they are very much central to the language and have a function.

Moreover, unlike directives, statements (except expression statements, of course, which are already addressed) are also not normally syntactically confusable with expressions by anyone who has some idea of the difference between the two, so these do not need to be called out as being permissible. Of the two possible exceptions:

1. A goofy block statement containing a label and statement (that without context is indistinguishable from a single-key object literal) perhaps also ought to be called out as not being a problem, but it is hardly likely that this will even occur as a possibility to most humans, and there is already an example (unfortunately without comment) indicating that an empty block statement, which can be confusable with an object literal, is not considered a problem.

2. An unused named function might be worth calling out as an example of something that will nevertheless be flagged as a problem when clearly used as an expression (since it can be confusable with a function declaration and people might wonder whether the linting code naively glosses over this distinction), but as its form as a declaration is well known as having utility in all environments, people will not need to see function declarations explicitly called out as not being problematic (though reference might still be made to the no-unused-vars rule).

Even for those who have read the spec, particularly due to the fact that "Hello world" is used as an example of an unused expression--without indicating that it is only considered as such if not in a prologue position (where it would then be a directive)--it may lead them to question whether we might be being loose in our definitions.

If an apartment contract states, "Allowable pets include birds", a conscientious tenant might still ask whether an ostrich or penguin is permissible, even if it is technically "obvious" that it is included since many landlords would not consider these as allowable, despite the "letter of the law". In many semantic concepts, there are often those consistent with typicality and those which are outliers (see https://en.wikipedia.org/wiki/Exemplar_theory ), and taking this into account helps us clarify things for users (and without inviting excessive questions). If using the spec terms alone were sufficient for the average person, we wouldn't need to have any examples at all.